### PR TITLE
RavenDB-16889 Index - `SelectMany` in nested dictionary

### PR DESCRIPTION
**Pull request by lastav5**
https://issues.hibernatingrhinos.com/issue/RavenDB-16889

### CODE REVIEW

File: test/SlowTests/Issues/RavenDB-16889.cs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        public RavenDB_16889(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        [Fact]
+        public void SelectManySimplePredicate()
+        {
+            TestCase<SelectManySimplePredicateIndex>();
+        }
+        public class SelectManySimplePredicateIndex : AbstractIndexCreationTask<TestObj>
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB-16889.cs]: private

File: test/SlowTests/Issues/RavenDB-16889.cs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+                Store("TagsResult", FieldStorage.Yes);
+            }
+        }
+
+        [Fact]
+        public void SelectManyComplexPredicate()
+        {
+            TestCase<SelectManyComplexPredicateIndex>();
+        }
+        public class SelectManyComplexPredicateIndex : AbstractIndexCreationTask<TestObj>
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB-16889.cs]: private

File: test/SlowTests/Issues/RavenDB-16889.cs [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+                Store("TagsResult", FieldStorage.Yes);
+            }
+        }
+
+        [Fact]
+        public void SelectManyComplexPredicate2Args()
+        {
+            TestCase<SelectManyComplexPredicate2ArgsIndex>();
+        }
+        public class SelectManyComplexPredicate2ArgsIndex : AbstractIndexCreationTask<TestObj>
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB-16889.cs]: private

File: test/SlowTests/Issues/RavenDB-16889.cs [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+            }
+        }
+
+        [Fact]
+        public void SelectManySimplePredicate2Args()
+        {
+            TestCase<SelectManySimplePredicate2ArgsIndex>();
+        }
+
+        public class SelectManySimplePredicate2ArgsIndex : AbstractIndexCreationTask<TestObj>
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB-16889.cs]: private

File: test/SlowTests/Issues/RavenDB-16889.cs [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+                    select new
+                    {
+                        TagsResult = taggable.Tags
+                            .SelectMany(v => v.Value, (pair, valuePair) => valuePair)
+                    };
+                Store("TagsResult", FieldStorage.Yes);
+            }
+        }
+        
+        public void TestCase<T>() where T : AbstractIndexCreationTask<TestObj>, new()
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB-16889.cs]: private

File: test/SlowTests/Issues/RavenDB-16889.cs [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+                else
+                {
+                    List<IDictionary> query = session.Query<Result, T>().Select(x => x.TagsResult).ToList();
+                    Assert.Equal(1, query.Count);
+                    Assert.Equal("value1", query[0]["key2"]);
+                }
+            }
+        }
+
+        public class TestObj
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB-16889.cs]: private

File: test/SlowTests/Issues/RavenDB-16889.cs [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+        {
+            public Dictionary<string, Dictionary<string, string>> Tags;
+        }
+
+        public class Result
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB-16889.cs]: private

File: test/SlowTests/Issues/RavenDB-16889.cs [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+            public IDictionary TagsResult;
+        }
+
+        [Fact]
+        public void UserTest()
+        {
+            TestCase<TestIndex>();
+        }
+
+        public class TestIndex : AbstractIndexCreationTask<TestObj>
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB-16889.cs]: private

File: test/SlowTests/Issues/RavenDB-16889.cs [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+                    {
+                        Count = taggable.Tags
+                            .SelectMany((v => v.Value))
+                            .Count()
+                    };
+                Store("Count", FieldStorage.Yes);
+            }
+        }
+
+        public class UserTestResult
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB-16889.cs]: private

File: test/SlowTests/Issues/RavenDB_16962.cs [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+                {
+                    var query = session.Query<SelectWithSimplePredicateIndex.Result, SelectWithSimplePredicateIndex>().ProjectInto<SelectWithSimplePredicateIndex.Result>().ToList();
+                    Assert.Equal(2, query.Count);
+                    Assert.Equal("FirstItem", query[0].Value);
+                    Assert.Equal("SecondItem", query[1].Value);
+                }
+            }
+        }
+
+        class SelectWithComplexPredicateIndex : AbstractIndexCreationTask<Response, SelectWithComplexPredicateIndex.Result>
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB_16962.cs]: private

File: test/SlowTests/Issues/RavenDB_16962.cs [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+                   {
+                       Value = item.Value,
+                       Index = item.Index
+                   };
+                Store("Value", FieldStorage.Yes);
+                Store("Index", FieldStorage.Yes);
+            }
+        }
+
+        class SelectWithSimplePredicateIndex : AbstractIndexCreationTask<Response, SelectWithSimplePredicateIndex.Result>
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB_16962.cs]: private

File: test/SlowTests/Issues/RavenDB_16962.cs [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+                    from item in response.Items.Select(x => new { Value = x.Value })
+                    select new Result
+                    {
+                        Value = item.Value
+                    };
+                Store("Value", FieldStorage.Yes);
+            }
+        }
+
+        class Item
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB_16962.cs]: private

File: test/SlowTests/Issues/RavenDB_16962.cs [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+        {
+            public string Value { get; set; }
+        }
+        class Response
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB_16962.cs]: private

**Review by ppekrol — APPROVED** [review]
Please check also formatting

File: test/SlowTests/Issues/RavenDB-16889.cs [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+            public Dictionary<string, Dictionary<string, string>> Tags;
+        }
+
+        private class Result
+        {
+            public IDictionary TagsResult;
+        }
+
+        [Fact]
+        private void UserTest()
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB-16889.cs]: tests should be public

Comment (lastav5) [reply test/SlowTests/Issues/RavenDB-16889.cs]: Missed this. Ty

**Comment by georgiosd** [comment]
@ppekrol Sorry to be a pain but I can't figure out whether this was merged into 5.2 also? Kindly point me in the right direction.

**Comment by ppekrol** [comment]
hi @georgiosd 

Yes, it was merged. We are always doing merges 'up' so from 4.2 to 5.1, from 5.1 to 5.2, etc. Best way is to click in the merge message of a PR (in this case `ppekrol merged commit c2cf29e into ravendb:v4.2 on Jul 28` ) and that https://github.com/ravendb/ravendb/commit/c2cf29e0c235270a033444e9e7a7fc092e8e9fd5
contains at the top all tags tag contain it: 
image

**Comment by georgiosd** [comment]
Thank you!

**Pull request merged**